Helpers/HttpExceptionFactory.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Exceptions\Helpers;
+
+use Exceptions\Http\Client;
+use Exceptions\Http\HttpExceptionInterface;
+use Exceptions\Http\Server;
+use InvalidArgumentException;
+use Throwable;
+
+class HttpExceptionFactory
+{
+    protected $mapping = [
+        400 => Client\BadRequestException::class,
+        401 => Client\UnauthorizedException::class,
+        402 => Client\PaymentRequiredException::class,
+        403 => Client\ForbiddenException::class,
+        404 => Client\NotFoundException::class,
+        405 => Client\MethodNotAllowedException::class,
+        406 => Client\NotAcceptableException::class,
+        407 => Client\ProxyAuthorizationRequiredException::class,
+        408 => Client\RequestTimeoutException::class,
+        409 => Client\ConflictException::class,
+        410 => Client\GoneException::class,
+        411 => Client\LengthRequiredException::class,
+        412 => Client\PreConditionRequiredException::class,
+        413 => Client\PayloadTooLargeException::class,
+        414 => Client\URITooLongException::class,
+        415 => Client\UnsupportedMediaTypeException::class,
+        416 => Client\RangeNotSatisfiableException::class,
+        417 => Client\ExpectationFailedException::class,
+        418 => Client\ImATeapotException::class,
+        421 => Client\MisdirectedRequestException::class,
+        422 => Client\UnprocessableEntityException::class,
+        423 => Client\LockedException::class,
+        424 => Client\FailedDependencyException::class,
+        426 => Client\UpgradeRequiredException::class,
+        428 => Client\PreConditionRequiredException::class,
+        429 => Client\TooManyRequestsException::class,
+        431 => Client\RequestHeaderFieldsTooLargeException::class,
+        451 => Client\UnavailableForLegalReasonsException::class,
+        500 => Server\InternalServerErrorException::class,
+        501 => Server\NotImplementedException::class,
+        502 => Server\BadGatewayException::class,
+        503 => Server\ServiceUnavailableException::class,
+        504 => Server\GatewayTimeoutException::class,
+        505 => Server\HttpVersionNotSupportedException::class,
+        507 => Server\InsuficientStorageException::class,
+        508 => Server\LoopDetectedException::class,
+    ];
+
+    /**
+     * @param int $responseCode
+     * @param string|null $message
+     * @param Throwable|null $ex
+     * @return HttpExceptionInterface
+     */
+    public function buildException(int $responseCode, string $message = '', Throwable $ex = null): HttpExceptionInterface
```

Comment:
I'd just name this `build`, it is already an `HttpExceptionFactory` so the `exception` part of the name here is redundant.

Reply:
Note that because this is a pure function, there isn't much reason not to put this static.

Reply:
Also, throwing an `InvalidArgumentException` is invalid here because the `InvalidArgumentException` is about type of data, not about content of data (or the domain), you should use the package's Data\ValidationException instead.

Reply:
> Note that because this is a pure function, there isn't much reason not to put this static.

In case of extending the factory to use custom pair of code/exception class, IMO it's easier to have non static property.
If it's static, you have to redeclare the whole` $mapping`; or maybe i'm wrong =/

Reply:
Then you should make the array accessible through a function that can be overriden, this allows people to call the parent and modify the return value.

Reply:
> Then you should make the array accessible through a function that can be overriden, this allows people to call the parent and modify the return value.

Still in static?

Reply:
You could move the array into the base static function that people call. Then once they want to override it, they can override the results by calling parent on it.

Reply:
Since the array is protected and no one will have a use for it, putting it into a static function makes perfect sense and makes that function become a data provider.

I prefer static in many cases when possible if it is pure functions and pure classes. It shows that you should not and will not be maintening state in the class which is good for testability, code reuse.

The only bad aspect would be the injection but since it is a single component that doesn't feature an interface, there is no point.